modules/image-loading-optimization/module.php

```diff
@@ -0,0 +1,346 @@
+		if ( ! str_contains( $image, ' width="' ) || ! str_contains( $image, ' height="' ) ) {
+			return $image;
+		}
+
+		$optimization_attrs = $this->get_loading_optimization_attributes(
+			array(
+				'width'         => $width,
+				'height'        => $height,
+				'loading'       => $loading_val,
+				'fetchpriority' => $fetchpriority_val,
+			)
+		);
+
+		if ( ! empty( $optimization_attrs['fetchpriority'] ) ) {
+			$image = str_replace( '<img', '<img fetchpriority="' . esc_attr( $optimization_attrs['fetchpriority'] ) . '"', $image );
+		}
+
+		if ( ! empty( $optimization_attrs['loading'] ) ) {
+			$image = str_replace( '<img', '<img loading="' . esc_attr( $optimization_attrs['loading'] ) . '"', $image );
+		}
+
+		return $image;
+	}
+
+	/**
+	 * Return loading Loading optimization attributes for a image with give attribute.
+	 *
+	 * @param array $attr Existing image attributes.
+	 * @return array Loading optimization attributes.
+	 */
+	private function get_loading_optimization_attributes( $attr ) {
+		$loading_attrs = array();
+
+		// For any resources, width and height must be provided, to avoid layout shifts.
+		if ( ! isset( $attr['width'], $attr['height'] ) ) {
+			return $loading_attrs;
+		}
+
+		/*
+		 * The key function logic starts here.
+		 */
+		$maybe_in_viewport    = null;
+		$increase_count       = false;
+		$maybe_increase_count = false;
+
+		// Logic to handle a `loading` attribute that is already provided.
+		if ( isset( $attr['loading'] ) ) {
+			/*
+			 * Interpret "lazy" as not in viewport. Any other value can be
+			 * interpreted as in viewport (realistically only "eager" or `false`
+			 * to force-omit the attribute are other potential values).
+			 */
+			if ( 'lazy' === $attr['loading'] ) {
+				$maybe_in_viewport = false;
+			} else {
+				$maybe_in_viewport = true;
+			}
+		}
+
+		// Logic to handle a `fetchpriority` attribute that is already provided.
+		if ( isset( $attr['fetchpriority'] ) && 'high' === $attr['fetchpriority'] ) {
+			/*
+			 * If the image was already determined to not be in the viewport (e.g.
+			 * from an already provided `loading` attribute), trigger a warning.
+			 * Otherwise, the value can be interpreted as in viewport, since only
+			 * the most important in-viewport image should have `fetchpriority` set
+			 * to "high".
+			 */
+			if ( false === $maybe_in_viewport ) {
+				_doing_it_wrong(
+					__FUNCTION__,
+					esc_html__( 'An image should not be lazy-loaded and marked as high priority at the same time.', 'elementor' ),
+					''
+				);
+				/*
+				 * Set `fetchpriority` here for backward-compatibility as we should
+				 * not override what a developer decided, even though it seems
+				 * incorrect.
+				 */
+				$loading_attrs['fetchpriority'] = 'high';
+			} else {
+				$maybe_in_viewport = true;
+			}
+		}
+
+		if ( null === $maybe_in_viewport ) {
+			//@todo: Only run this functionality for frontend like in wp core.
```

Comment:
why are we adding todo? who is going to do it and when? is there a task?

Reply:
@TzviRabinovitch I think if this todo is a must - we should create the task. @kt-12 please let me know if it's relevant

Reply:
Good catch. I'll take another look, but I believe this was to work around an issue with the Unit test suite always running in the admin context, which was causing an early return here.

Reply:
Addressed in [6e26a95](https://github.com/elementor/elementor/pull/23631/commits/6e26a955605265ae922790bdece4f1e4afc75689)

Reply:
@joemcgill If I understand correctly, should this logic only run in the front end? If so, can you check if `Plugin::$instance->editor->is_edit_mode()` works better?